pkg/migrations/sql/idempotent/tracing-public.sql

```diff
@@ -0,0 +1,459 @@
+
+CREATE OR REPLACE FUNCTION SCHEMA_TRACING_PUBLIC.span_tree(_trace_id SCHEMA_TRACING_PUBLIC.trace_id, _span_id bigint, _max_dist int default null)
+RETURNS TABLE
+(
+    trace_id SCHEMA_TRACING_PUBLIC.trace_id,
+    parent_span_id bigint,
+    span_id bigint,
+    dist int,
+    idx int,
+    path bigint[]
+)
+AS $func$
+    SELECT
+        trace_id,
+        parent_span_id,
+        span_id,
+        dist * -1 as dist,
+        idx,
+        path
+    FROM SCHEMA_TRACING_PUBLIC.upstream_spans(_trace_id, _span_id, _max_dist)
+    UNION ALL
+    SELECT
+        trace_id,
+        parent_span_id,
+        span_id,
+        dist,
+        idx,
+        path
+    FROM SCHEMA_TRACING_PUBLIC.downstream_spans(_trace_id, _span_id, _max_dist) d
+    WHERE d.dist != 0
+$func$ LANGUAGE sql STABLE PARALLEL SAFE;
+GRANT EXECUTE ON FUNCTION SCHEMA_TRACING_PUBLIC.span_tree(SCHEMA_TRACING_PUBLIC.trace_id, bigint, int) TO prom_reader;
+
+CREATE OR REPLACE FUNCTION SCHEMA_TRACING_PUBLIC.put_tag_key(_key SCHEMA_TRACING_PUBLIC.tag_k, _tag_type SCHEMA_TRACING_PUBLIC.tag_type)
+RETURNS VOID
+AS $func$
+    INSERT INTO SCHEMA_TRACING.tag_key AS k (key, tag_type)
+    VALUES (_key, _tag_type)
+    ON CONFLICT (key) DO
+    UPDATE SET tag_type = k.tag_type | EXCLUDED.tag_type
+    WHERE k.tag_type & EXCLUDED.tag_type = 0
+$func$
+LANGUAGE SQL VOLATILE STRICT;
+GRANT EXECUTE ON FUNCTION SCHEMA_TRACING_PUBLIC.put_tag_key(SCHEMA_TRACING_PUBLIC.tag_k, SCHEMA_TRACING_PUBLIC.tag_type) TO prom_writer;
+
+CREATE OR REPLACE FUNCTION SCHEMA_TRACING_PUBLIC.put_tag(_key SCHEMA_TRACING_PUBLIC.tag_k, _value SCHEMA_TRACING_PUBLIC.tag_v, _tag_type SCHEMA_TRACING_PUBLIC.tag_type)
+RETURNS VOID
+AS $func$
+    INSERT INTO SCHEMA_TRACING.tag AS a (tag_type, key_id, key, value)
+    SELECT _tag_type, ak.id, _key, _value
+    FROM SCHEMA_TRACING.tag_key ak
+    WHERE ak.key = _key
+    ON CONFLICT (key, value) DO
+    UPDATE SET tag_type = a.tag_type | EXCLUDED.tag_type
+    WHERE a.tag_type & EXCLUDED.tag_type = 0
+$func$
+LANGUAGE SQL VOLATILE STRICT;
+GRANT EXECUTE ON FUNCTION SCHEMA_TRACING_PUBLIC.put_tag(SCHEMA_TRACING_PUBLIC.tag_k, SCHEMA_TRACING_PUBLIC.tag_v, SCHEMA_TRACING_PUBLIC.tag_type) TO prom_writer;
+
+CREATE OR REPLACE FUNCTION SCHEMA_TRACING_PUBLIC.jsonb(_tag_map SCHEMA_TRACING_PUBLIC.tag_map)
+RETURNS jsonb
+AS $func$
+    /*
+    takes an tag_map which is a map of tag_key.id to tag.id
+    and returns a jsonb object containing the key value pairs of tags
+    */
+    SELECT jsonb_object_agg(a.key, a.value)
+    FROM jsonb_each(_tag_map) x -- key is tag_key.id, value is tag.id
+    INNER JOIN LATERAL -- inner join lateral enables partition elimination at execution time
+    (
+        SELECT
+            a.key,
+            a.value
+        FROM SCHEMA_TRACING.tag a
+        WHERE a.id = x.value::text::bigint
+        -- filter on a.key to eliminate all but one partition of the tag table
```

Comment:
On ingest, we have the key, but not necessarily the key_id (unless we have it in a cache somewhere).
When we do something like `span_tags ? ('http.status_code' !== 200)` we are starting out with the key, not the id.

I think in more cases, partitioning on key will be better, but I could be convinced otherwise.